shell/renderer/api/atom_api_context_bridge.cc

```diff
@@ -0,0 +1,473 @@
+    return v8::MaybeLocal<v8::Value>(handle);
+  }
+
+  // Errors aren't serializable currently, we need to pull the message out and
+  // re-construct in the destination context
+  if (value->IsNativeError()) {
+    v8::Context::Scope destination_context_scope(destination);
+    return v8::MaybeLocal<v8::Value>(v8::Exception::Error(
+        v8::Exception::CreateMessage(destination->GetIsolate(), value)->Get()));
+  }
+
+  // Manually go through the array and pass each value individually into a new
+  // array so that functions deep insidee arrays get proxied or arrays of
+  // promises are proxied correctly.
+  if (IsPlainArray(value)) {
+    v8::Context::Scope destination_context_scope(destination);
+    v8::Local<v8::Array> arr = v8::Local<v8::Array>::Cast(value);
+    size_t length = arr->Length();
+    v8::Local<v8::Array> cloned_arr =
+        v8::Array::New(destination->GetIsolate(), length);
+    for (size_t i = 0; i < length; i++) {
+      auto value_for_array = PassValueToOtherContext(
+          source, destination, arr->Get(source, i).ToLocalChecked(), store,
+          recursion_depth + 1);
+      if (value_for_array.IsEmpty())
+        return v8::MaybeLocal<v8::Value>();
+
+      if (!mate::internal::IsTrue(
+              cloned_arr->Set(destination, static_cast<int>(i),
+                              value_for_array.ToLocalChecked()))) {
+        return v8::MaybeLocal<v8::Value>();
+      }
+    }
+    store->CacheProxiedObject(value, cloned_arr);
+    return v8::MaybeLocal<v8::Value>(cloned_arr);
+  }
+
+  // Proxy all objects
+  if (IsPlainObject(value)) {
+    auto object_value = v8::Local<v8::Object>::Cast(value);
+    auto passed_value = CreateProxyForAPI(object_value, source, destination,
+                                          store, recursion_depth + 1);
+    if (passed_value.IsEmpty())
+      return v8::MaybeLocal<v8::Value>();
+    return v8::MaybeLocal<v8::Value>(passed_value.ToLocalChecked());
+  }
+
+  // Serializable objects
+  blink::CloneableMessage ret;
+  {
+    v8::Context::Scope source_context_scope(source);
+    // V8 serializer will throw an error if required
+    if (!mate::ConvertFromV8(source->GetIsolate(), value, &ret))
+      return v8::MaybeLocal<v8::Value>();
+  }
+
+  v8::Context::Scope destination_context_scope(destination);
+  return v8::MaybeLocal<v8::Value>(
+      mate::ConvertToV8(destination->GetIsolate(), ret));
+}
+
+v8::Local<v8::Value> ProxyFunctionWrapper(
+    context_bridge::RenderFramePersistenceStore* store,
+    size_t func_id,
+    mate::Arguments* args) {
+  // Context the proxy function was called from
+  v8::Local<v8::Context> calling_context = args->isolate()->GetCurrentContext();
+  // Context the function was created in
+  v8::Local<v8::Context> func_owning_context =
+      std::get<1>(store->functions()[func_id]).Get(args->isolate());
+
+  v8::Context::Scope func_owning_context_scope(func_owning_context);
+  v8::Local<v8::Function> func =
```

Comment:
its better to block scope to avoid any compiler optimizations here since we clearly want this call to execute after entering a context scope.

```
v8::Context::Scope scope();
{
  // scope related operatios.
}
```

also in other places in this PR.